### PR TITLE
Send "Name" and "ID" when stating stopped containers

### PR DESCRIPTION
**Pull request by WeiZhang555**
When `docker stats` stopped containers, client will get empty stats data,
this commit will gurantee client always get "Name" and "ID" field, so
that it can format with `ID` and `Name` fields successfully.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

<del>
This commit fixes panic when execute stats command:

* use `--format {{.Name}}` with `--all` when there're exited containers.
* use `--format {{.Name}}` while stating exited container.

The root cause is when stating an exited container, the result from the
api didn't contain the `Name` and `ID` field, which will make `format`
process panic.

Panic log is like this:

```
panic: runtime error: slice bounds out of range [recovered]
	panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0xb20f80, 0xc420014110)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
text/template.errRecover(0xc4201773e8)
	/usr/local/go/src/text/template/exec.go:140 +0x2ad
panic(0xb20f80, 0xc420014110)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/docker/docker/cli/command/formatter.(*containerStatsContext).Name(0xc420430160,
0x0, 0x0)
	/go/src/github.com/docker/docker/cli/command/formatter/stats.go:148
+0x86
reflect.Value.call(0xb9a3a0, 0xc420430160, 0x2213, 0xbe3657, 0x4,
0x11bc9f8, 0x0, 0x0, 0x4d75b3, 0x1198940, ...)
	/usr/local/go/src/reflect/value.go:434 +0x5c8
reflect.Value.Call(0xb9a3a0, 0xc420430160, 0x2213, 0x11bc9f8, 0x0, 0x0,
0xc420424028, 0xb, 0xb)
	/usr/local/go/src/reflect/value.go:302 +0xa4
text/template.(*state).evalCall(0xc420177368, 0xb9a3a0, 0xc420430160,
0x16, 0xb9a3a0, 0xc420430160, 0x2213, 0x1178fa0, 0xc4203ea330,
0xc4203de283, ...)
	/usr/local/go/src/text/template/exec.go:658 +0x530
```

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>
</del>

Introduced in https://github.com/docker/docker/pull/27797, ping @vdemeester

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]
LGTM 🐻
If we could add at least a test in `cli/command/formatter` for `stats` that would be really cool 👼

**Comment by WeiZhang555** [comment]
Added an integration test, let me see if I can add a unit test too :smile:

**Comment by WeiZhang555** [comment]
Unit tests are also added now :smile:

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

File: daemon/stats.go
Comment (coolljt0725) [inline daemon/stats.go]: why not return `nil` here?

File: daemon/stats.go
Comment (LK4D4) [inline daemon/stats.go]: sound more like `time.Ticker` for me

**Review by LK4D4 — CHANGES_REQUESTED** [review]

File: daemon/stats.go
Comment (WeiZhang555) [inline daemon/stats.go]: We can't.

If `--no-stream` is set, client will fetch the stat data once and only once then quit, if we return nil here, client will never get the data, so we need to `continue` to try again to send the real data.

Besides, I use `continue` here is also for interrupting next step of sending empty stats to client, but try one more again to send the real data. 

Integration test `TestStatsAllRunningNoStream` is for this use case.

File: daemon/stats.go
Comment (WeiZhang555) [inline daemon/stats.go]: What's the difference? I don't see the big advantage over `time.After()` :smile:

Comment (WeiZhang555) [inline daemon/stats.go]: According to explaination: https://golang.org/pkg/time/#Ticker , I don't think `Ticker is what I want, It will send the time periodically, what I want is a `timeout` mechanism.

**Comment by WeiZhang555** [comment]
I increased the timeout a little bit: `case <-time.After(statsInterval + 100*time.Millisecond):`, to make sure `case v, ok := <-updates:` always goes first. It's not big change, so we don't need Vincent's re-LGTM :smile:

**Review by vdemeester — APPROVED** [review]
> It's not big change, so we don't need Vincent's re-LGTM :smile: 

:scream_cat: :smirk_cat: 

still LGTM 🐻

**Comment by WeiZhang555** [comment]
What's current status? ping @LK4D4 @coolljt0725 @cpuguy83

File: cli/command/formatter/stats.go
Comment (thaJeztah) [inline cli/command/formatter/stats.go]: Alternatively;

```go
return strings.TrimPrefix(c.s.Name, "/")
```

Comment (thaJeztah) [inline cli/command/formatter/stats.go]: I was discussing this with @vdemeester - we should probably have a general function for doing this (out of scope for this PR).

Given that the current implementation is closest to the original code, I think it's ok to keep it as-is (`TrimPrefix` will also be a slightly heavier operation)

Comment (WeiZhang555) [inline cli/command/formatter/stats.go]: So are you ok with this PR ? :smile:

**Comment by WeiZhang555** [comment]
ping @thaJeztah @LK4D4 @vieux

File: cli/command/formatter/stats.go
Comment (AkihiroSuda) [inline cli/command/formatter/stats.go]: I wonder this should be `"--"` https://github.com/docker/docker/pull/30281/files#diff-02748098eadf6b07cde6c7d8f8bbc9c7R149

cc @yongtang

Comment (WeiZhang555) [inline cli/command/formatter/stats.go]: This change is only for avoiding panic, in my opinion, we should fix this on daemon side, we should pass the name of container to client in all situations. Then `c.s.Name` will never be empty.

To me, "--" or "" are the same, they both represent the wrong cases.

Comment (AkihiroSuda) [inline cli/command/formatter/stats.go]: Makes sense, but for consistency (see CPUPerc(), MemUsage(), ...), I thought `"--"` seems better.
But this is nitpicky and overall LGTM.

Comment (yongtang) [inline cli/command/formatter/stats.go]: There are actually two panic cases:

In case there are only dead containers, `containerStatsContext` has values in certain fields, but not name and ID. (`IsInvalid` is `true`)

In case there is no active containers, the panic is caused by faking of the header:
https://github.com/docker/docker/blob/04f7a03359ac009ad2cd548fa93124a66e880ae5/cli/command/formatter/formatter.go#L74-L78
In this situation, the `containerStatsContext` simply has no value and every field is default empty value. (even `IsInvalid` is `false` because false is default.)

For that we do need to return some value here as `c.s.Name` could be empty.

Comment (WeiZhang555) [inline cli/command/formatter/stats.go]: I'm fine with "--", already updated :smile:

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
UT failing

```
00:07:22.339 --- FAIL: TestContainerStatsContextWrite (0.00s)
00:07:22.339 	assert.go:107: stats_test.go:127: Expected 'container1  abcdef  foo
00:07:22.340 		container2    
00:07:22.340 		' (string) got 'container1  abcdef  foo
00:07:22.340 		container2    --
00:07:22.340 		' (string)
00:07:22.340 FAIL
```

**Comment by WeiZhang555** [comment]
@AkihiroSuda forgot to update my unit test, sorry. Now it should be fine. :smile:

File: cli/command/formatter/stats.go
Comment (LK4D4) [inline cli/command/formatter/stats.go]: We should eliminate those `/` at all. They have no real value now. I'm ok with slicing.

File: daemon/stats.go
Comment (LK4D4) [inline daemon/stats.go]: I don't really understand this trick with timeout. Can you elaborate?

**Comment by WeiZhang555** [comment]
@LK4D4 To make myself clearer, here is the scenario:

 Suppose we are running `docker stats --format "table {{.ID}}\t{{.CPUPerc}}\t{{.Name}}" 33979dba3f1` and `33979dba3f1` is a stopped container with a name `foo`, what we expected is result like:

```
CONTAINER ID                                                        CPU %            NAME
33979dba3f1173451820ffc06cf25c06abbf1c061cd42f6d05beba1c1b518659    --               clever_snyder
```

Right? Currently it will panic, so the client side change is for fixing the panic.

As to the daemon side change (for explaining you comments), is for fixing another problem, for non-running container, the `case v, ok := <-updates:` will never return, so the client will get nothing. After client side timeout, it will print something like this:

```
CONTAINER ID    CPU %    NAME
                 --               
```

IMO, it's reasonable that we can't get the resources, so "--" is correct for "CPU %" field, but "ID" and "NAME" shouldn't be empty, even if it's stopped container, we should be able to get its ID and Name. 
**Hence** adding a timeout will guarantee that client can always get "ID" and "Name" even for stopped containers.

What do you think?

**Comment by LK4D4** [comment]
@WeiZhang555 I don't think that daemon-side fix is correct. Not sure how it works at all with `all` and `stream`, but even without those concerns, "wait for 100 ms" is a quite shady strategy in go scheduler world. I believe you can construct something around `WaitWithContext` from `container/state.go` to wait on container stop in select. Feel free to ping me in IRC or mail if you need help with that.
Thanks!

**Comment by WeiZhang555** [comment]
@LK4D4 Actually I think you don't need to worry so much about the `100 ms`, in this current implementation, it's `case <-time.After(statsInterval + 100*time.Millisecond):` which means the timeout period is `1.1s`, in my previous implementation, it's `case <-time.After(statsInterval):` (1s), they both works(see https://github.com/docker/docker/pull/29702#issuecomment-269435236), and the timeout period doesn't influence, it's only a guarantee that daemon will return something before client timeout(2s). 

The reason I choose `1.1s` is that client timeout is 2s, and daemon's stats collector has a collection period of 1s, so I choose a middle number to make sure client can get data(especially "ID" and "Name") before timeout. You don't need to care too much about `1.1s` or `1.2s` or `1.25s`, they are the same, even the golang's timer isn't accurate enough,it still doesn't influence.

From tomorrow it's Chinese's traditional New Year, so I will be on vacation for the next 10 days. I'll chat with you about your idea on IRC or mail after I'm back.

Happy Chinese New Year, **EVERYONE**! :smile: :fireworks:

**Comment by thaJeztah** [comment]
Happy Chinese New Year to you, your family and colleagues, @WeiZhang555!

**Comment by LK4D4** [comment]
@WeiZhang555 I think it might be not the best decision to base daemon code on client behavior(timeout) because Docker client is not the only user of API. I'll see what we can do. Happy Chinese New Year in the meantime.

**Comment by LK4D4** [comment]
`all` is not supposed to show stopped containers at all - that's a bug.
There is check in `statsCollector.run`:
```
				if _, ok := err.(errNotRunning); !ok {
					logrus.Errorf("collecting stats for %s: %v", pair.container.ID, err)
				}
				continue

```
I think we just need to publish bogus stats there.
I dunno who was author of that, maybe @crosbymichael or @mlaventure have opinions.

**Comment by thaJeztah** [comment]
@LK4D4 isn't that what was changed in https://github.com/docker/docker/pull/20835 ?

**Comment by LK4D4** [comment]
@thaJeztah I actually don't know :/ I'm quite confused about client-daemon relationships here.
ping @cpuguy83

**Comment by cpuguy83** [comment]
I think `--all` is supposed to show all containers running or not running.
`docker stats` with no args show all containers which are running and adds/removes entries from the list as stuff goes up/down.

**Comment by LK4D4** [comment]
@cpuguy83 I think there is events watcher in CLI which removes watch from stopped containers. I don't really want to read that code :)
However, my comment about putting empty stats for  stopped containers looks correct.

**Comment by mlaventure** [comment]
There's no stats for non running containers, showing 0 values for those would make sense to me (as per LK4D4 suggestion)

**Comment by LK4D4** [comment]
@WeiZhang555 let's create separate PR for cli fix and then discuss daemon change here.

**Comment by WeiZhang555** [comment]
Client fix is moved to https://github.com/docker/docker/pull/30776 , let's focus on daemon side fix here.

**Comment by WeiZhang555** [comment]
@LK4D4 How about this one?

**Comment by WeiZhang555** [comment]
ping @LK4D4

**Review by LK4D4 — APPROVED** [review]
LGTM

**Comment by LK4D4** [comment]
ping @docker/core-engine-maintainers

**Review by mlaventure — APPROVED** [review]
LGTM

**Comment by mlaventure** [comment]
ping @thaJeztah does that need any doc updates?

**Comment by mstanleyjones** [comment]
We'll need to update all the docs that show the return value for running `docker stats` on stopped containers. I suspect it's not much if any. Please raise a [docs issue](https://github.com/docker/docker.github.io/issues) and include the relevant version where this change will land.

**Comment by LK4D4** [comment]
@WeiZhang555 can you take care of docs issue? Thanks!

**Comment by WeiZhang555** [comment]
OK, I'll make a follow up PR for docs :wink:

**Pull request merged**
